ddl/ddl_api.go

```diff
@@ -6298,6 +6390,124 @@ func (d *ddl) AlterTablePartitionAttributes(ctx sessionctx.Context, ident ast.Id
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Followers = op.UintValue
+			case ast.TableOptionPlacementVoterCount:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Voters = op.UintValue
+			case ast.TableOptionPlacementLearnerCount:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Learners = op.UintValue
+			case ast.TableOptionPlacementSchedule:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Schedule = op.StrValue
+			case ast.TableOptionPlacementConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Constraints = op.StrValue
+			case ast.TableOptionPlacementLeaderConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.LeaderConstraints = op.StrValue
+			case ast.TableOptionPlacementLearnerConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.LearnerConstraints = op.StrValue
+			case ast.TableOptionPlacementFollowerConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.FollowerConstraints = op.StrValue
+			case ast.TableOptionPlacementVoterConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.VoterConstraints = op.StrValue
+			}
+		}
+	}
+
+	// Can not use both a placement policy and direct assignment. If you alter specify both in a CREATE TABLE or ALTER TABLE an error will be returned.
+	if ptDef.DirectPlacementOpts != nil && ptDef.PlacementPolicyRef != nil {
+		return errors.Trace(ErrPlacementPolicyWithDirectOption.GenWithStackByArgs(ptDef.PlacementPolicyRef.Name))
+	}
+	if ptDef.DirectPlacementOpts != nil {
+		// check the direct placement option compatibility.
+		if err := checkPolicyValidation(ptDef.DirectPlacementOpts); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	if ptDef.PlacementPolicyRef != nil {
+		// placement policy reference will override the direct placement options.
+		policy, ok := ctx.GetInfoSchema().(infoschema.InfoSchema).PolicyByName(ptDef.PlacementPolicyRef.Name)
+		if !ok {
+			return errors.Trace(infoschema.ErrPlacementPolicyNotExists.GenWithStackByArgs(ptDef.PlacementPolicyRef.Name))
+		}
+		ptDef.PlacementPolicyRef.ID = policy.ID
+	}
+	job := &model.Job{
+		SchemaID:   schema.ID,
+		TableID:    meta.ID,
+		SchemaName: schema.Name.L,
+		Type:       model.ActionAlterTablePartitionOptions,
```

Comment:
As talked in the parser, it is too big as an action type. We actually only process policy options for table in this ddl job. And you may find it difficult to process other options.